import sys
import time
import grpc
import random
from datetime import datetime

import weather_pb2 as pb
import weather_pb2_grpc as pb_grpc

def readings_generator():
    # ØªÙˆÙ„ÙŠØ¯ Ù‚ÙŠÙ… Ø¹Ø´ÙˆØ§Ø¦ÙŠØ© Ù„ÙƒÙ„ Ù‚Ø±Ø§Ø¡Ø©
    for i in range(10):  # 10 Ù‚Ø±Ø§Ø¡Ø§Øª
        # Ù‚ÙŠÙ… Ø¹Ø´ÙˆØ§Ø¦ÙŠØ© ÙˆØ§Ù‚Ø¹ÙŠØ© Ù„Ù„Ø·Ù‚Ø³
        temperature = random.uniform(-10, 45)    # Ù…Ù† -10 Ø¥Ù„Ù‰ 45 Ù…Ø¦ÙˆÙŠØ©
        humidity = random.uniform(20, 95)        # Ù…Ù† 20% Ø¥Ù„Ù‰ 95% Ø±Ø·ÙˆØ¨Ø©
        pressure = random.uniform(980, 1040)     # Ù…Ù† 980 Ø¥Ù„Ù‰ 1040 Ù‡ÙŠÙƒØªÙˆØ¨Ø§Ø³ÙƒØ§Ù„
        
        r = pb.SensorReading(
            sensor_id=f"python-sensor-{random.randint(1000, 9999)}",
            temperature=round(temperature, 2),
            humidity=round(humidity, 2),
            pressure=round(pressure, 2),
            ts=int(time.time() * 1000),
        )
        
        current_time = datetime.now().strftime("%H:%M:%S")
        print(f"[{current_time}] ğŸ“¤ Ø¥Ø±Ø³Ø§Ù„ Ù‚Ø±Ø§Ø¡Ø© #{i+1}:")
        print(f"   ğŸ†” Ø§Ù„Ù…Ø³ØªØ´Ø¹Ø±: {r.sensor_id}")
        print(f"   ğŸŒ¡ï¸  Ø§Ù„Ø­Ø±Ø§Ø±Ø©: {r.temperature:.2f}Â°C")
        print(f"   ğŸ’§ Ø§Ù„Ø±Ø·ÙˆØ¨Ø©: {r.humidity:.2f}%")
        print(f"   ğŸ“Š Ø§Ù„Ø¶ØºØ·: {r.pressure:.2f} hPa")
        print("   " + "â”€" * 30)
        
        yield r
        time.sleep(1)  # Ø§Ù†ØªØ¸Ø± Ø«Ø§Ù†ÙŠØ© Ø¨ÙŠÙ† ÙƒÙ„ Ù‚Ø±Ø§Ø¡Ø©

def main():
    host = sys.argv[1] if len(sys.argv) > 1 else "192.168.96.202"  # ØªØ£ÙƒØ¯ Ù…Ù† Ø§Ù„Ø¹Ù†ÙˆØ§Ù†
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 50052
    target = f"{host}:{port}"
    
    print("=" * 50)
    print("ğŸŒ¤ï¸  Ø¹Ù…ÙŠÙ„ Ù…Ø³ØªØ´Ø¹Ø± Ø§Ù„Ø·Ù‚Ø³ - Ø¨ÙŠØ§Ù†Ø§Øª Ø¹Ø´ÙˆØ§Ø¦ÙŠØ©")
    print("=" * 50)
    print(f"ğŸ“ Ø§Ù„Ù‡Ø¯Ù: {target}")
    print(f"ğŸ”„ Ø¥Ø±Ø³Ø§Ù„ 10 Ù‚Ø±Ø§Ø¡Ø§Øª Ø¹Ø´ÙˆØ§Ø¦ÙŠØ©...")
    print("=" * 50)
    
    # Ø¥Ø¶Ø§ÙØ© metadata Ù„Ù„ØªÙˆØ«ÙŠÙ‚
    metadata = [('authorization', 'python-token-12345')]
    
    try:
        with grpc.insecure_channel(target) as channel:
            stub = pb_grpc.SensorServiceStub(channel)
            # Ø¥Ø±Ø³Ø§Ù„ Ù…Ø¹ metadata
            ack = stub.StreamSensorData(readings_generator(), metadata=metadata)
            print(f"âœ… ØªÙ… Ø§Ù„Ø§Ø³ØªÙ„Ø§Ù…: ok={ack.ok}, msg='{ack.msg}'")
    except grpc.RpcError as e:
        print(f"âŒ Ø®Ø·Ø£ ÙÙŠ Ø§Ù„Ø§ØªØµØ§Ù„: {e}")
    except Exception as e:
        print(f"âŒ Ø®Ø·Ø£ ØºÙŠØ± Ù…ØªÙˆÙ‚Ø¹: {e}")

if __name__ == '__main__':
    main()